src/hssm/param/params.py

```diff
@@ -0,0 +1,353 @@
+        if param_name in kwargs:
+            if param_name in user_params:
+                raise ValueError(
+                    f"Parameter {param_name} specified in both `include` and `kwargs`."
+                )
+            user_params[param_name] = UserParam.from_kwargs(
+                param_name,
+                # Use pop to remove the key from kwargs
+                kwargs.pop(param_name),
+            )
+
+    # Process p_outliers the same way.
+    if model.has_lapse:
+        user_params["p_outlier"] = UserParam.from_kwargs("p_outlier", p_outlier)
+
+    return user_params
+
+
+def make_params(model: HSSM, user_params: dict[str, UserParam]) -> dict[str, Param]:
+    """Make parameters from a dict of UserParams.
+
+    Parameters
+    ----------
+    model
+        The HSSM model.
+    user_params
+        A dictionary with the parameter names as keys and UserParam objects as values.
+
+    Returns
+    -------
+    dict[str, Param]
+        A dictionary with the parameter names as keys and Param objects as values.
+    """
+    params = {}
+    is_ddm = (
+        model.model_name in ["ddm", "ddm_sdv", "full_ddm"]
+        and model.loglik_kind != "approx_differentiable"
+    )
+
+    for name in model.list_params:
+        if name in user_params:
+            param: Param = make_param_from_user_param(model, name, user_params[name])
+        else:
+            param = make_param_from_defaults(model, name)
+
+        if model.prior_settings == "safe":
+            if isinstance(param, RegressionParam):
+                param.make_safe_priors(model.data, model.additional_namespace, is_ddm)
+
+        param.process_prior()
+        params[name] = param
+
+    return params
+
+
+def make_param_from_user_param(model: HSSM, name: str, user_param: UserParam) -> Param:
+    """Create a Param object from a UserParam object.
+
+    Parameters
+    ----------
+    model
+        The HSSM model.
+    name
+        The name of the parameter.
+    user_param
+        A UserParam object.
+
+    Returns
+    -------
+    Param
+        A Param object with the specified parameters.
+    """
+    default_prior, default_bounds = model.model_config.get_defaults(name)
+
+    # Use information in the UserParam object to determine if the parameter is a
+    # regression parameter or a simple parameter
+    if user_param.is_regression:
+        param: Param = RegressionParam.from_user_param(user_param)
+    elif user_param.is_simple:
+        param = SimpleParam.from_user_param(user_param)
+    else:
+        # If we cannot tell definitively that the parameter is a regression parameter
+        # or a simple parameter from the UserParam object,
+        # we need to determine the type based on whether the model has a global formula
+        if model.global_formula is not None:
+            param = RegressionParam.from_user_param(user_param)
+        else:
+            param = SimpleParam.from_user_param(user_param)
```

Comment:
Why not add the type hints here?

Reply:
Again, the type mypy uses is when it first sees the `param` variable in the first branch. Without the type hint, mypy will use `RegressionParam` for `param`, which will cause errors. So once I declare `param` to be `Param`, that works for all branches